### PR TITLE
Add DFU support to the main application.

### PR DESCRIPTION
**Pull request by nomis52**

### CODE REVIEW

**Comment by peternewman** [comment]
Fix Travis

File: firmware/ja-rule.X/nbproject/configurations.xml [hunk 1]
```diff
@@ -203,6 +211,9 @@
   <sourceRootList>
     <Elem>../src</Elem>
     <Elem>../src/system_config/mx_795_512l</Elem>
+    <Elem>../../../blink/linker</Elem>
```
Comment (peternewman) [inline firmware/ja-rule.X/nbproject/configurations.xml]: ?

File: firmware/src/system_config/mx_795_512l/system_init.c [hunk 2]
```diff
@@ -131,7 +134,7 @@ const USB_DEVICE_FUNCTION_REGISTRATION_TABLE funcRegistrationTable[2] =
         .driver = (void*)USB_DEVICE_CDC_FUNCTION_DRIVER,    /* USB CDC function data exposed to device layer */
         .funcDriverInit = (void*)&cdcInit0    /* Function driver init data */
   },
-  /* Function 2 */
+  /* Function 2 - The Ja Rule Interface*/
```
Comment (peternewman) [inline firmware/src/system_config/mx_795_512l/system_init.c]: Add a space after interface

File: firmware/src/system_config/mx_795_512l/system_init.c [hunk 3]
```diff
@@ -333,24 +353,42 @@ sd001 = {
...
+struct {
+  uint8_t bLength;
+  uint8_t bDscType;
+  uint16_t string[UID_LENGTH * 2 + 1];
+}
+
+serial_number_string_descriptor = {
+  sizeof(serial_number_string_descriptor),
+  USB_DESCRIPTOR_STRING, {
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
```
Comment (peternewman) [inline firmware/src/system_config/mx_795_512l/system_init.c]: UID?

Comment (peternewman) [inline firmware/src/system_config/mx_795_512l/system_init.c]: Ah, just read above.

File: firmware/src/usb_transport.c [hunk 4]
```diff
@@ -112,20 +131,37 @@ void USBTransport_EventHandler(USB_DEVICE_EVENT event, void* event_data,
...
     case USB_DEVICE_EVENT_CONTROL_TRANSFER_SETUP_REQUEST:
       // This means we have received a setup packet
-      setupPacket = (USB_SETUP_PACKET*) event_data;
-      if (setupPacket->bRequest == USB_REQUEST_SET_INTERFACE) {
+      setup_packet = (USB_SETUP_PACKET*) event_data;
+      if (setup_packet->RequestType == USB_SETUP_REQUEST_TYPE_CLASS &&
+          setup_packet->Recipient == USB_SETUP_REQUEST_RECIPIENT_INTERFACE &&
+          setup_packet->DataDir == USB_SETUP_REQUEST_DIRECTION_HOST_TO_DEVICE &&
+          setup_packet->bRequest == DFU_DETATCH &&
+          setup_packet->wIndex == 3u) {
```
Comment (peternewman) [inline firmware/src/usb_transport.c]: Make this a constant?

**Comment by peternewman** [comment]
Generally LGTM

File: firmware/src/usb_transport.c (hunk 4 above)
Comment (nomis52) [inline firmware/src/usb_transport.c]: Done

File: firmware/src/system_config/mx_795_512l/system_init.c (hunk 3 above)
Comment (nomis52) [inline firmware/src/system_config/mx_795_512l/system_init.c]: Yup

File: firmware/src/system_config/mx_795_512l/system_init.c (hunk 2 above)
Comment (nomis52) [inline firmware/src/system_config/mx_795_512l/system_init.c]: Added

File: firmware/ja-rule.X/nbproject/configurations.xml (hunk 1 above)
Comment (nomis52) [inline firmware/ja-rule.X/nbproject/configurations.xml]: Removed

**Comment by peternewman** [comment]
LGTM

**Pull request merged**